File: tools/parity/runner.cxx
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Simon Geard
//
// Parity runner — C++.  Usage: runner_cpp <script> <result-file>
// Builds the canonical parity graph (see GRAPH.md), run_file the script with
// echo on, writes the normalised structured trailer to <result-file>.

#include "cmdgraph.hxx"

#include <fstream>
#include <iostream>
#include <string>

using namespace cmdgraph;

static ActionResult act_echo(const ArgList& a, const std::string&) {
    std::cout << "echo: " << arg_str(a[0]) << '\n';
    return action_ok();
}
static ActionResult act_add(const ArgList& a, const std::string&) {
    std::cout << "sum: " << (arg_int(a[0]) + arg_int(a[1])) << '\n';
    return action_ok();
}
static ActionResult act_save(const ArgList&, const std::string&) {
    std::cout << "save: ok\n";
    return action_ok();
}
static ActionResult act_send(const ArgList&, const std::string&) {
    std::cout << "send: ok\n";
    return action_ok();
}
static ActionResult act_scale(const ArgList&, const std::string&) {
    std::cout << "scale: ok\n";
    return action_ok();
}
static ActionResult act_open(const ArgList& a, const std::string&) {
    int id = arg_int(a[0]);
    if (id <= 0) return action_ok();
    return action_ok(std::to_string(id));
}
static ActionResult act_zero(const ArgList&, const std::string&) {
    return action_ok("0");
}
static ActionResult act_where(const ArgList&, const std::string& ctx) {
    std::cout << "where: ctx=" << ctx << '\n';
    return action_ok();
}
static ActionResult act_update(const ArgList& a, const std::string&) {
    std::cout << "update: " << arg_str(a[0]) << '\n';
    return action_ok();
}
static void enter_detail(const std::string& ctx) {
    std::cout << "entered detail ctx=" << ctx << '\n';
}

int main(int argc, char** argv) {
    if (argc < 3) {
        std::cerr << "usage: runner_cpp <script> <result-file>\n";
        return 2;
    }
    const std::string script  = argv[1];
    const std::string resfile = argv[2];

    Engine eng;
    eng.add_state("root", "root> ");
    eng.add_command("root", "e(cho)",  EdgeKind::Action,
                    {.proc=act_echo, .help="echo text", .args={arg_is_rest("text")}});
    eng.add_command("root", "ad(d)",   EdgeKind::Action,
                    {.proc=act_add, .help="add two ints",
                     .args={arg_is_int("x"), arg_is_int("y")}});
    eng.add_command("root", "s(ave)",  EdgeKind::Action, {.proc=act_save, .help="save"});
    eng.add_command("root", "s(end)",  EdgeKind::Action, {.proc=act_send, .help="send"});
    eng.add_command("root", "sc(ale)", EdgeKind::Action,
                    {.proc=act_scale, .help="scale", .args={arg_is_real("f")}});
    eng.add_command("root", "o(pen)",  EdgeKind::DoGoto,
                    {.target="detail", .proc=act_open, .help="open id",
                     .args={arg_is_int("id")}});
    eng.add_command("root", "z(ero)",  EdgeKind::DoGoto,
                    {.target="detail", .proc=act_zero, .help="zero-ctx do_goto"});
    eng.add_command("root", "g(o)",    EdgeKind::Goto,  {.target="detail", .help="go"});
    eng.add_command("root", "q(uit)",  EdgeKind::Quit,  {.help="quit"});

    eng.add_state("detail", "detail> ");
    eng.set_on_enter("detail", enter_detail);
    eng.add_command("detail", "w(here)",  EdgeKind::Action,
                    {.proc=act_where, .help="show context"});
    eng.add_command("detail", "u(pdate)", EdgeKind::DoPop,
                    {.proc=act_update, .help="update note", .args={arg_is_rest("note")}});
    eng.add_command("detail", "b(ack)",   EdgeKind::Pop,  {.help="back"});
    eng.add_command("detail", "q(uit)",   EdgeKind::Quit, {.help="quit"});

    eng.finalize("root");

    RC  st   = RC::Ok;
    int line = 0;
    bool ok  = eng.run_file(script, true, &st, &line);

    std::string rc;
    if (!ok && line == 0) {
        rc = "OPEN_FAIL";
    } else if (ok) {
        rc = "OK";
    } else {
        switch (st) {
            case RC::Unknown:      rc = "UNKNOWN";      break;
            case RC::Ambiguous:    rc = "AMBIGUOUS";    break;
            case RC::Transitioned: rc = "TRANSITIONED"; break;
            case RC::Exited:       rc = "EXITED";       break;
            case RC::Error:        rc = "ERROR";        break;
            default:               rc = "OK";           break;
        }
    }

    std::ofstream f(resfile);
    f << "ok=" << (ok ? 1 : 0) << '\n';
    f << "rc=" << rc << '\n';
    f << "line=" << line << '\n';
    f << "state=" << eng.current_state() << '\n';
    f << "last_message=" << eng.last_message << '\n';
    f << "last_error=" << eng.last_error << '\n';
    return 0;
}

File: tools/parity/runner.f90
!! SPDX-License-Identifier: MIT
!! Copyright (c) 2026 Simon Geard
!!
! Parity runner — Fortran.  Usage: runner_f <script> <result-file>
! Builds the canonical parity graph (see GRAPH.md), run_file the script with
! echo on, writes the normalised structured trailer to <result-file>.

module parity_procs
    use cmdgraph
    use dlist
    implicit none
contains

    function arg_i(args, n) result(v)
        type(dlist_t), intent(in)             :: args
        integer, intent(in)                   :: n
        integer                               :: v
        class(dlist_node_data_t), allocatable :: nd
        v = 0
        nd = args%get(n)
        select type (nd)
        type is (dlist_node_integer); v = nd%data
        end select
    end function arg_i

    function arg_s(args, n) result(v)
        type(dlist_t), intent(in)             :: args
        integer, intent(in)                   :: n
        character(len=:), allocatable         :: v
        class(dlist_node_data_t), allocatable :: nd
        v = ""
        nd = args%get(n)
        select type (nd)
        type is (dlist_node_char); v = nd%data
        end select
    end function arg_s

    function act_echo(args, ctx) result(rv)
        type(dlist_t), intent(in)    :: args
        character(len=*), intent(in) :: ctx
        type(action_result_t)        :: rv
        write(*,'(2a)') "echo: ", arg_s(args, 1)
        rv = action_ok()
    end function act_echo

    function act_add(args, ctx) result(rv)
        type(dlist_t), intent(in)    :: args
        character(len=*), intent(in) :: ctx
        type(action_result_t)        :: rv
        write(*,'(a,i0)') "sum: ", arg_i(args, 1) + arg_i(args, 2)
        rv = action_ok()
    end function act_add

    function act_save(args, ctx) result(rv)
        type(dlist_t), intent(in)    :: args
        character(len=*), intent(in) :: ctx
        type(action_result_t)        :: rv
        write(*,'(a)') "save: ok"
        rv = action_ok()
    end function act_save

    function act_send(args, ctx) result(rv)
        type(dlist_t), intent(in)    :: args
        character(len=*), intent(in) :: ctx
        type(action_result_t)        :: rv
        write(*,'(a)') "send: ok"
        rv = action_ok()
    end function act_send

    function act_scale(args, ctx) result(rv)
        type(dlist_t), intent(in)    :: args
        character(len=*), intent(in) :: ctx
        type(action_result_t)        :: rv
        write(*,'(a)') "scale: ok"
        rv = action_ok()
    end function act_scale

    function act_open(args, ctx) result(rv)
        type(dlist_t), intent(in)    :: args
        character(len=*), intent(in) :: ctx
        type(action_result_t)        :: rv
        integer                      :: id
        character(len=16)            :: buf
        id = arg_i(args, 1)
        if (id <= 0) then
            rv = action_ok()
        else
            write(buf,'(i0)') id
            rv = action_ok(trim(buf))
        end if
    end function act_open

    function act_zero(args, ctx) result(rv)
        type(dlist_t), intent(in)    :: args
        character(len=*), intent(in) :: ctx
        type(action_result_t)        :: rv
        rv = action_ok("0")
    end function act_zero

    function act_where(args, ctx) result(rv)
        type(dlist_t), intent(in)    :: args
        character(len=*), intent(in) :: ctx
        type(action_result_t)        :: rv
        write(*,'(2a)') "where: ctx=", ctx
        rv = action_ok()
    end function act_where

    function act_update(args, ctx) result(rv)
        type(dlist_t), intent(in)    :: args
        character(len=*), intent(in) :: ctx
        type(action_result_t)        :: rv
        write(*,'(2a)') "update: ", arg_s(args, 1)
        rv = action_ok()
    end function act_update

    subroutine enter_detail(ctx)
        character(len=*), intent(in) :: ctx
        write(*,'(2a)') "entered detail ctx=", ctx
    end subroutine enter_detail

end module parity_procs


program runner_f
    use cmdgraph
    use parity_procs
    implicit none

    type(engine_t)                :: eng
    character(len=4096)           :: script, resfile
    integer                       :: st, ln, u, nargs
    logical                       :: ok
    character(len=:), allocatable :: em, rc, lmsg, lerr

    nargs = command_argument_count()
    if (nargs < 2) then
        write(*,'(a)') "usage: runner_f <script> <result-file>"
        stop 2
    end if
    call get_command_argument(1, script)
    call get_command_argument(2, resfile)

    call eng%add_state("root", prompt="root> ")
    call eng%add_command("root", "e(cho)",  EDGE_ACTION,  proc=act_echo, &
                         help="echo text", args=[arg_is_rest("text")])
    call eng%add_command("root", "ad(d)",   EDGE_ACTION,  proc=act_add, &
                         help="add two ints", args=[arg_is_int("x"), arg_is_int("y")])
    call eng%add_command("root", "s(ave)",  EDGE_ACTION,  proc=act_save, help="save")
    call eng%add_command("root", "s(end)",  EDGE_ACTION,  proc=act_send, help="send")
    call eng%add_command("root", "sc(ale)", EDGE_ACTION,  proc=act_scale, &
                         help="scale", args=[arg_is_real("f")])
    call eng%add_command("root", "o(pen)",  EDGE_DO_GOTO, target="detail", &
                         proc=act_open, help="open id", args=[arg_is_int("id")])
    call eng%add_command("root", "z(ero)",  EDGE_DO_GOTO, target="detail", &
                         proc=act_zero, help="zero-ctx do_goto")
    call eng%add_command("root", "g(o)",    EDGE_GOTO,    target="detail", help="go")
    call eng%add_command("root", "q(uit)",  EDGE_QUIT,    help="quit")

    call eng%add_state("detail", prompt="detail> ")
    call eng%set_on_enter("detail", enter_detail)
    call eng%add_command("detail", "w(here)",  EDGE_ACTION, proc=act_where, &
                         help="show context")
    call eng%add_command("detail", "u(pdate)", EDGE_DO_POP, proc=act_update, &
                         help="update note", args=[arg_is_rest("note")])
    call eng%add_command("detail", "b(ack)",   EDGE_POP,    help="back")
    call eng%add_command("detail", "q(uit)",   EDGE_QUIT,   help="quit")

    call eng%finalize("root")

    st = RC_OK
    ln = 0
    ok = eng%run_file(trim(script), echo=.true., stat=st, errmsg=em, line=ln)

    if ((.not. ok) .and. ln == 0) then
        rc = "OPEN_FAIL"
    else if (ok) then
        rc = "OK"
    else
        select case (st)
        case (RC_UNKNOWN);      rc = "UNKNOWN"
        case (RC_AMBIGUOUS);    rc = "AMBIGUOUS"
        case (RC_TRANSITIONED); rc = "TRANSITIONED"
        case (RC_EXITED);       rc = "EXITED"
        case (RC_ERROR);        rc = "ERROR"
        case default;           rc = "OK"
        end select
    end if

    lmsg = ""
    if (allocated(eng%last_message)) lmsg = eng%last_message
    lerr = ""
    if (allocated(eng%last_error))   lerr = eng%last_error

    open(newunit=u, file=trim(resfile), status="replace", action="write")
    write(u,'(a,i0)') "ok=", merge(1, 0, ok)
    write(u,'(2a)')   "rc=", rc
    write(u,'(a,i0)') "line=", ln
    write(u,'(2a)')   "state=", eng%current_state()
    write(u,'(2a)')   "last_message=", lmsg
    write(u,'(2a)')   "last_error=", lerr
    close(u)
end program runner_f
